### PR TITLE
#167727477 Create Facilities And Rooms

### PR DESCRIPTION
**Pull request by chiomadans1759**
### What does this PR do?
- Creates the Facilities and rooms endpoints that enables travel admin list their accommodation facilities and also add rooms to those facilities

### Description of tasks to be completed
`For the Facilities Endpoint`
- Create the Facilities endpoint
- Create validations for a facility input fields
- Create a middleware that ensure all facility fields are properly inputed
- Protect the endpoint to be accessed only by Super Admins and Travel Admins

`For the Rooms Endpoint`
- Create the Rooms endpoint
- Create validations for a room input fields
- Create a middleware that ensure all room fields are properly inputed
- Protect the endpoint to be accessed only by Super Admins and Travel Admins

### How should this be manually tested?
-  Clone this repository to your local machine
-  Run `npm install` to install node packages
-  On your Api Testing Tool, make a `POST` request to `http://localhost:3000/api/v1/facilities` to create facilities.
-  On your Api Testing Tool, make a `POST` request to `http://localhost:3000/api/v1/facilities/rooms` to add rooms to a facility
- Run `npm test` to test endpoint

### Sample Request Objects
`For Facilities`
```
{
  "name": "Kings Palace Hotel",
  "address": "Otolo, Nnewi",
  "number_of_rooms": 8,
  "available_space": 7,
  "images": ["https://cloudinary.com/16x16/1.png","https://cloudinary.com/16x16/1.png"],
  "description": "Our services are the best"
}
```

`For Rooms`
```
{
  "name": "5B",
  "type": "Executive",
  "price": 8000,
  "facility_id": 2,
  "images": ["https://cloudinary.com/16x16/1.png", "https://cloudinary.com/16x16/1.png"]
}
```

### What are the relevant Pivotal Tacker stories?
[finishes #167727477](https://www.pivotaltracker.com/story/show/167727477)

### Who should take a look
@okoroemeka 

### Screenshot
Screen Shot 2019-09-05 at 8 50 41 AM
Screen Shot 2019-09-05 at 8 51 25 AM

### CODE REVIEW

**Comment by parkerthegeniuschild** [comment]
[x] How are you sure your user has `Postman` installed? You might reword it as: `Use an API Testing Tool e.g Postman, Apigee, Insomnia, Paw, Rest Assured etc`
---
image
---
[x] Are we to make a `PUT, PATCH or GET Request?`
---
[x] What should the `request body contain`? `profile_id, date_of_birth, color_of_car or zodiac_sign?`
---

**Comment by jiokeokwuosa** [comment]
Hi, @chiomadans1759 Are we to upload image while testing the facility endpoint? it will be helpful if the fields to be populated is listed out

**Comment by chiomadans1759** [comment]
> ## [x] How are you sure your user has `Postman` installed? You might reword it as: `Use an API Testing Tool e.g Postman, Apigee, Insomnia, Paw, Rest Assured etc`
> ## image
> ## [x] Are we to make a `PUT, PATCH or GET Request?`
> ## [x] What should the `request body contain`? `profile_id, date_of_birth, color_of_car or zodiac_sign?`

You are right @parkerthegeniuschild. Thanks for pointing that out. I have effected the feedback

**Comment by chiomadans1759** [comment]
> Hi, @chiomadans1759 Are we to upload image while testing the facility endpoint? it will be helpful if the fields to be populated is listed out

The Guide on the PR description has been updated. Kindly check

**Comment by jiokeokwuosa** [comment]
Ok. i have tested both endpoint locally its working just fine, kindly note that some reviewers will encounter difficulties testing the endpoint because of the nature of the images field input, it worked well when i set postman to raw -> json. you can as well specify that in your PR to help other reviewers. i will go ahead and approve

**Review by jiokeokwuosa — APPROVED** [review]

**Comment by chiomadans1759** [comment]
> Ok. i have tested both endpoint locally its working just fine, kindly note that some reviewers will encounter difficulties testing the endpoint because of the nature of the images field input, it worked well when i set postman to raw -> json. you can as well specify that in your PR to help other reviewers. i will go ahead and approve

I have dropped a post on the group on how to properly pass an array as part of x-www-form-urlencoded. Hopefully, that should help. Cheers

**Comment by parkerthegeniuschild** [comment]
[x] Images MUST BE an ARRAY
---
image

Imagine, trying to bulk-insert a CSV file containing 1,000 links to images, do I have to first convert them into an array myself? No support for CSV values? I feel you could have taken a string OR `a string containing comma-separated items e.g image1, image2, image3`. Then split them at the back-end to form an array of strings, before insertion. This way, if a user tries to upload CSVs containing tons of images, they'll be happy with you.

**Comment by chiomadans1759** [comment]
> ## [x] Images MUST BE an ARRAY
> image
> 
> Imagine, trying to bulk-insert a CSV file containing 1,000 links to images, do I have to first convert them into an array myself? No support for CSV values? I feel you could have taken a string OR `a string containing comma-separated items e.g image1, image2, image3`. Then split them at the back-end to form an array of strings, before insertion. This way, if a user tries to upload CSVs containing tons of images, they'll be happy with you.

We are trying to build an MVP @parkerthegeniuschild A user will have to insert the images as an array for now. That is what the story specified. Thanks

**Comment by parkerthegeniuschild** [comment]
> > ## [x] Images MUST BE an ARRAY
> > image
> > Imagine, trying to bulk-insert a CSV file containing 1,000 links to images, do I have to first convert them into an array myself? No support for CSV values? I feel you could have taken a string OR `a string containing comma-separated items e.g image1, image2, image3`. Then split them at the back-end to form an array of strings, before insertion. This way, if a user tries to upload CSVs containing tons of images, they'll be happy with you.
> 
> We are trying to build an MVP @parkerthegeniuschild A user will have to insert the images as an array for now. That is what the story specified. Thanks

The issue of `MVP` does not tally with `working without the correct input`. You're depending on your user to convert the input into an array at the front-end. Meanwhile, your controller should actually be doing the conversion at the back-end. And you're the same person that said, `stories were written by humans`; we should go through it and make up for what is lost. Why you're always contradicting yourself.?

**Comment by chiomadans1759** [comment]
> > > ## [x] Images MUST BE an ARRAY
> > > image
> > > Imagine, trying to bulk-insert a CSV file containing 1,000 links to images, do I have to first convert them into an array myself? No support for CSV values? I feel you could have taken a string OR `a string containing comma-separated items e.g image1, image2, image3`. Then split them at the back-end to form an array of strings, before insertion. This way, if a user tries to upload CSVs containing tons of images, they'll be happy with you.
> > 
> > 
> > We are trying to build an MVP @parkerthegeniuschild A user will have to insert the images as an array for now. That is what the story specified. Thanks
> 
> The issue of `MVP` does not tally with `working without the correct input`. You're depending on your user to convert the input into an array at the front-end. Meanwhile, your controller should actually be doing the conversion at the back-end. And you're the same person that said, `stories were written by humans`; we should go through it and make up for what is lost. Why you're always contradicting yourself.?

I will look into this and find a possible fix for this @parkerthegeniuschild. Thanks for your feedback

File: src/validation/facilities.validation.js [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+    {
+      field: error.param,
+      message: error.msg
+    }
+  )
+});
+
+
+const validateFacilities = [
+  check('name', 'Kindly Provide a name for your facility')
```
Comment (Qausim) [inline src/validation/facilities.validation.js]: Can you ensure name, address, description are strings? Putting a number (note, not in quotes) in there passes through? Address and description should at least have letters or a mixture of numbers and letters

File: src/validation/rooms.validation.js [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  formatter: (error) => (
+    {
+      field: error.param,
+      message: error.msg
+    }
+  )
+});
+
+const validateRooms = [
+  check('name', 'Kindly Provide a name for your room')
```
Comment (Qausim) [inline src/validation/rooms.validation.js]: `name` here should have letters too (like "Room 5", as saying just "5" won't make much sense.

File: src/validation/rooms.validation.js [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    .not()
+    .isEmpty()
+    .isLength({ min: 2 })
+    .withMessage('Too short: enter a minimum of 2 characters'),
+
+  check('type', 'Kindly enter the type of room')
```
Comment (Qausim) [inline src/validation/rooms.validation.js]: Same for `type`

**Review by Qausim — CHANGES_REQUESTED** [review]

File: src/validation/facilities.validation.js (hunk 1 above)
Comment (chiomadans1759) [reply src/validation/facilities.validation.js]: Alright. I will Implement this. Thanks

File: src/validation/rooms.validation.js (hunk 2 above)
Comment (chiomadans1759) [reply src/validation/rooms.validation.js]: About this, we can't really influence how people will want to name the rooms on their facilities. I will ensure that the name is a string but I don't think it's wise to force them to follow a naming convention. Thanks for the feedback

**Review by Qausim — COMMENTED** [review]
Cool

**Pull request closed without merging**